ouroboros-consensus-cardano/tools/db-analyser/Analysis.hs

```diff
@@ -0,0 +1,210 @@
+  HasAnalysis
+-------------------------------------------------------------------------------}
+
+class HasAnalysis blk where
+    countTxOutputs   :: blk -> Int
+    blockHeaderSize  :: blk -> Natural
+    blockTxs         :: blk -> [BL.ByteString]
+    mkTopLevelConfig :: [FilePath] -- Genesis file or files.
+                     -> Bool       -- is it mainnet?
+                     -> IO (TopLevelConfig blk)
+
+{-------------------------------------------------------------------------------
+  ByronBlock instance
+-------------------------------------------------------------------------------}
+
+instance HasAnalysis ByronBlock where
+    countTxOutputs = aBlockOrBoundary (const 0) countTxOutputsByron
+    blockHeaderSize =
+      aBlockOrBoundary blockBoundaryHeaderSize blockHeaderSizeByron
+    blockTxs = aBlockOrBoundary (const []) blockTxsByron
+    mkTopLevelConfig [configFile] onMainNet = do
+      genesisConfig <- openGenesisByron configFile onMainNet
+      return $ mkByronTopLevelConfig genesisConfig
+    mkTopLevelConfig ls _onMainNet =
+      error $
+        "a single genesis file is needed for pure Byron. Given " ++ show ls
+
+-- | Equivalent of 'either' for 'ABlockOrBoundary'.
+aBlockOrBoundary :: (Chain.ABoundaryBlock ByteString -> a)
+                 -> (Chain.ABlock ByteString -> a)
+                 -> ByronBlock -> a
+aBlockOrBoundary fromBoundary fromRegular blk = case blk of
+    Byron.ByronBlock (Chain.ABOBBoundary boundaryBlock) _ _
+      -> fromBoundary boundaryBlock
+    Byron.ByronBlock (Chain.ABOBBlock regularBlk) _ _
+      -> fromRegular regularBlk
+
+countTxOutputsByron :: Chain.ABlock ByteString -> Int
+countTxOutputsByron Chain.ABlock{..} = countTxPayload bodyTxPayload
+  where
+    Chain.AHeader{..} = blockHeader
+    Chain.ABody{..}   = blockBody
+
+    countTxPayload :: Chain.ATxPayload a -> Int
+    countTxPayload = sum
+                   . map (countTx . unAnnotated . Chain.aTaTx)
+                   . Chain.aUnTxPayload
+
+    countTx :: Chain.Tx -> Int
+    countTx = length . Chain.txOutputs
+
+blockBoundaryHeaderSize ::  Chain.ABoundaryBlock ByteString -> Natural
+blockBoundaryHeaderSize =
+    fromIntegral . BS.length . Chain.boundaryHeaderAnnotation . Chain.boundaryHeader
+
+blockHeaderSizeByron ::  Chain.ABlock ByteString -> Natural
+blockHeaderSizeByron = Chain.headerLength . Chain.blockHeader
+
+blockTxsByron :: Chain.ABlock ByteString -> [BL.ByteString]
+blockTxsByron block = map (BL.fromStrict . Chain.aTaAnnotation) blockTxAuxs
+  where
+    Chain.ABlock{ blockBody } = block
+    Chain.ABody{ bodyTxPayload } = blockBody
+    Chain.ATxPayload{ aUnTxPayload = blockTxAuxs } = bodyTxPayload
+
+openGenesisByron :: FilePath -> Bool -> IO Genesis.Config
+openGenesisByron configFile onMainNet = do
+    genesisHash <- either (error . show) return =<< runExceptT
+      (snd <$> Genesis.readGenesisData configFile)
+    Right genesisConfig <- runExceptT $
```

Comment:
I thought you used `error` to avoid that incomplete pattern match on `Right genesisHash`, but you still do that for `genesisConfig`? I would do the same thing for both.